database/misc/mongodb_replication.py

```diff
@@ -0,0 +1,378 @@
+#
+
+def main():
+    module = AnsibleModule(
+        argument_spec = dict(
+            login_user=dict(default=None),
+            login_password=dict(default=None),
+            login_host=dict(default='localhost'),
+            login_port=dict(default='27017'),
+            replica_set=dict(default=None),
+            host_name=dict(default='localhost'),
+            host_port=dict(default='27017'),
+            host_type=dict(default='replica', choices=['replica','arbiter']),
+            ssl=dict(default=False),
+            build_indexes = dict(type='bool', choices=BOOLEANS, default='yes'),
+            hidden = dict(type='bool', choices=BOOLEANS, default='no'),
+            priority = dict(default='1.0'),
+            slave_delay = dict(type='int', default='0'),
+            votes = dict(type='int', default='1'),
+            state=dict(default='present', choices=['absent', 'present']),
+        )
+    )
+
+    if not pymongo_found:
+        module.fail_json(msg='the python pymongo (>= 2.4) module is required')
+
+    login_user = module.params['login_user']
+    login_password = module.params['login_password']
+    login_host = module.params['login_host']
+    login_port = module.params['login_port']
+    replica_set = module.params['replica_set']
+    host_name = module.params['host_name']
+    host_port = module.params['host_port']
+    host_type = module.params['host_type']
+    ssl = module.params['ssl']
+    state = module.params['state']
+
+    replica_set_created = False
+
+    try:
+        if replica_set is None:
+            module.fail_json(msg='replica_set parameter is required')
+        else:
+            client = MongoReplicaSetClient(login_host, int(login_port), replicaSet=replica_set, ssl=ssl)
+
+        authenticate(client, login_user, login_password)
+
+    except ConnectionFailure, e:
+        module.fail_json(msg='unable to connect to database: %s' % str(e))
+    except ConfigurationError:
+        try:
+            client = MongoClient(login_host, int(login_port), ssl=ssl)
+            authenticate(client, login_user, login_password)
+            if state == 'present':
+                config = { '_id': "{0}".format(replica_set), 'members': [{ '_id': 0, 'host': "{0}:{1}".format(host_name, host_port)}] }
+                client['admin'].command('replSetInitiate', config)
+                wait_for_ok_and_master(module, client)
+                replica_set_created = True
+                module.exit_json(changed=True, host_name=host_name, host_port=host_port, host_type=host_type)
+        except OperationFailure, e:
+            module.fail_json(msg='Unable to initiate replica set: %s' % str(e))
+
+    check_members(state, module, client, host_name, host_port, host_type)
+
+    if state == 'present':
+        if host_name is None and not replica_set_created:
+            module.fail_json(msg='host_name parameter required when adding new host into replica set')
+
+        try:
+            if not replica_set_created:
```

Comment:
actually i check_members isn't enough - as it is, you can't change a replica's vote count, priority, hidden flag, etc. As long as it already exists in the repset check_members will find it and exit the module. I think it's better to do these checks in the specific functions.